### PR TITLE
Hypothesis for parameters

### PR DESCRIPTION
**Pull request by AdrianSosic**
This PR introduces hypothesis strategies for the parameters package and rewrites the parameter testing logic:

- The previous hand-selected set of serialization tests is replaced with hypothesis-generated test scenarios, covering a much broader range.
- Bugs encountered during this process were fixed and the parameter validation logic was correspondingly extended.
- A new set of "validation" tests is now responsible for the scenarios that are not covered by hypothesis (i.e. exception tests for invalid parameter specifications). These replace the previous incomplete set of parameter tests and now thoroughly cover all parameter validation steps by triggering the expected errors in a systematic way.

### CODE REVIEW

File: baybe/parameters/custom.py [hunk 1]
```diff
@@ -54,21 +56,30 @@ def _validate_custom_data(  # noqa: DOC101, DOC103
...
+        if value.select_dtypes("number").shape[1] != value.shape[1]:
             raise ValueError(
-                f"The custom dataframe for parameter {self.name} contains NaN "
-                f"entries, which is not supported."
+                f"The custom dataframe for parameter {self.name} contains "
+                f"non-numeric values."
             )
-        if len(value) != len(set(value.index)):
+        if not all(isinstance(x, str) for x in value.index):
             raise ValueError(
```
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply baybe/parameters/custom.py:65]: Thanks, good catch, forgot about the docstring.... Added it now! Will push the changes once I've addressed all your comments.

File: tests/hypothesis_strategies.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+parameter_name = st.text(min_size=1)
+"""A strategy that generates parameter names."""
+
+categories = st.lists(st.text(min_size=1), min_size=2, unique=True)
+"""A strategy that generates parameter categories."""
+
+
+@st.composite
+def smiles(draw: st.DrawFn):
+    """Generate short SMILES strings."""
```
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply tests/hypothesis_strategies.py:46]: For the hypothesis part, I don't think this would be critical. Basically similar to my response to Alex when he asked about whether it's worth having more complex molecules: we only want to create something for our backend to digest that is close to the real use case.

However, for the backend part, this is worth a discussion. What if the user entered two smiles that are equivalent? Perhaps it's worth doing the additional check there by going via `canonical`?

So what I'm trying to say: for generation we don't necessarily need to bother, but for validation we perhaps should!? And if we do, THEN we would of course need to adjust the hypothesis strategy as well. 

Do you think this would cause some major delays in parameter creation? I guess our caching might save us here as well. Could you point me to that function, or even better, implement the check yourself, since you are the expert here?

File: tests/hypothesis_strategies.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+
+@st.composite
+def custom_encodings(draw: st.DrawFn):
+    """Generate data for class:`baybe.parameters.custom.CustomDiscreteParameter`."""
+    index = st.lists(st.text(min_size=1), min_size=2, max_size=10, unique=True)
+    cols = columns(
+        names_or_number=10,
+        elements=st.floats(allow_nan=False, allow_infinity=False),
+        unique=True,
```
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply tests/hypothesis_strategies.py:70]: Not exactly sure if I understand your point.... These are all keyword args, so there should be no whitespace?! Or am I looking at the wrong spot?

File: baybe/parameters/numerical.py [hunk 4]
```diff
@@ -54,12 +56,14 @@ def _validate_tolerance(  # noqa: DOC101, DOC103
...
             ValueError: If the tolerance is not safe.
         """
-        # NOTE: computing all pairwise distances can be avoided if we ensure that the
-        #   values are ordered (which is currently not the case)
-        dists = pdist(np.asarray(self.values).reshape(-1, 1))
-        max_tol = dists.min() / 2.0
+        min_dist = np.diff(self.values).min()
+        if min_dist == (eps := np.nextafter(0, 1, dtype=DTypeFloatNumpy)):
+            raise NumericalUnderflowError(
+                f"The distance between any two parameter values must be at least "
```
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply baybe/parameters/numerical.py:62]: Hm, perhaps the case `tolerance=0.0` is special in the sense that we wouldn't even need to carry out the validation in the first place? Because the only requirement for `tolerance=0.0` is really that all values are distinct, which is however already asserted by the unique value validator. As soon as we have non-zero tolerance though, I run into the potential floating point underflow due to the `/ 2.0` division and couldn't quickly find a better alternative 🤔 So what do you suggest? Would it suffice to have a `if tolerance == 0.0` guard clause that skips validation in that case? Or do we need to polish the underflow logic?

File: tests/serialization/test_parameter_serialization.py [hunk 5]
```diff
@@ -1,24 +1,15 @@
 """Test serialization of parameters."""
```
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply tests/serialization/test_parameter_serialization.py:1]: Yes, it's the former. In fact, I also thought about this issue already before but decided not to bother too much :D But now that you also bring it up, I guess this is valid point. Have a look at my new version that I'll push in a few mins. I guess you will like it a lot better 👍🏼

[the remark this answers is not in the document]
Comment (AdrianSosic) [reply tests/serialization/test_parameter_serialization.py:1]: PS: I would still keep the `parameters` strategy in the strategies file, because I think it will turn out to be useful in other test contexts, where you simply need "some random parameter" ... But it's no longer used for the serialization test in my new version

File: tests/hypothesis_strategies.py [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+@st.composite
+def substance_data(draw: st.DrawFn):
+    """Generate data for class:`baybe.parameters.substance.SubstanceParameter`."""
+    names = draw(st.lists(st.text(min_size=1), min_size=2, max_size=10, unique=True))
+    substances = draw(st.lists(smiles(), min_size=len(names), max_size=len(names)))
+    return dict(zip(names, substances))
+
+
+@st.composite
+def custom_encodings(draw: st.DrawFn):
```
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply tests/hypothesis_strategies.py:64]: Valid point, have changed accordingly

File: tests/hypothesis_strategies.py [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+    upper = draw(st.floats(min_value=lower, exclude_min=True, allow_infinity=False))
+    return NumericalContinuousParameter(name=name, bounds=(lower, upper))
+
+
+@st.composite
+def categorical_parameter(draw: st.DrawFn):
+    """Generate class:`baybe.parameters.categorical.CategoricalParameter`."""
+    name = draw(parameter_name)
+    values = draw(categories)
+    encoding = draw(st.sampled_from(CategoricalEncoding))
```
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply tests/hypothesis_strategies.py:114]: For the strategies I would say "no". Because their only task is to span the space of possible objects, which is already done with current version. Adding the string option only makes the strategy unnecessarily complicated (there is also the `shrinkage` concept of hypothesis that you need to keep in mind) without changing the output at all.

We could debate about whether we tests like this could be added as additional validation/initialization tests. However, I not sure if this particular already crosses the line of what still belongs to our framework and what is outside of it. In a sense, your suggested test would test the attrs converter, not our parameter object. If you follow that logic to the end, it would mean that we would need to add tests for all input types that our converters accept. That is, if we have `converter=list`, your suggestion would translate into having explicit test cases for all object types that can be converter into lists, i.e. basically everything that implements the `typing.Sequence` interface.

**TL;DR:** I consider the particular enum case we are talking about as an edge case. If you think it should be included, we could add outside hypothesis, but I would not add it the the strategy.

File: tests/hypothesis_strategies.py [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+                st.floats(allow_infinity=False, allow_nan=False),
+            ),
+            min_size=2,
+            unique=True,
+        )
+    )
+
+    # Reject examples where the tolerance validator cannot be satisfied
+    try:
+        return NumericalDiscreteParameter(name=name, values=values)
```
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply tests/hypothesis_strategies.py:95]: Yes and no, would appreciate your thoughts here:

So far, I deliberately did **not** add it, because providing a **valid** choice would require copying the same logic that is conducted **inside** the parameter class, introducing unnecessary redundancy and coupling. The current approach simply says "let the class itself decide what is appropriate", which is very robust and maintainable. Also, this type of rejection sampling is a core principle of hypothesis itself. 

A different reason why we might want to include it though (i.e. not only to pass validation) is code coverage. If we think that different tolerance values should be tested, then we should change the code. Here, we then run into the question if we want to get appropriate values by rejection sampling (current approach) or by replicating the acceptance-logic to draw valid values in the first place.

File: baybe/parameters/custom.py [hunk 9]
```diff
@@ -54,21 +56,30 @@ def _validate_custom_data(  # noqa: DOC101, DOC103
...
-                f"The custom dataframe for parameter {self.name} contains "
-                f"duplicated indices. Please only provide dataframes with unique"
-                f" indices."
+                f"The custom dataframe for parameter {self.name} contains non-string "
+                f"index values."
             )
-        if value.select_dtypes("number").shape[1] != value.shape[1]:
+        if not all(len(x) > 0 for x in value.index):
+            raise ValueError(
+                f"The custom dataframe for parameter {self.name} contains empty string "
```
[the remark this answers is not in the document]
Comment (AVHopp) [reply baybe/parameters/custom.py:72]: Why didn't the linters complain? I guess because the docstring already contained a `ValueError`?

File: tests/hypothesis_strategies.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (AVHopp) [reply tests/hypothesis_strategies.py:46]: If the user enters two SMILES that describe the same molecule, this is a duplicate input, since what the user wants to use as a parameter is probably the *molecule described by the SMILES*, and not the SMILES. So in my opinion, this needs to be handled like every other part of the code in which we prevent the user from making multiple input (that is, unless there is any reason to have the same molecule twice with different representations *on purpose*, but I cannot think of this case).

So in my opinion, we need to either verify that a SMILES is canonical and provide a warning if we detect that the provided SMILES is not canonical or make everything canonical by ourselves. Since this is only a minor check (and should only be done once, right?!), I do not think that this creates an immense overhead.

File: tests/hypothesis_strategies.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (AVHopp) [reply tests/hypothesis_strategies.py:70]: I think that @Scienfitz wants e.g. `names_or_number = 10` instead of `names_or_number=10`. I am however not aware of any regulations/rules which variant is preferable.

File: baybe/parameters/categorical.py [hunk 10]
```diff
@@ -22,10 +22,13 @@ class CategoricalParameter(DiscreteParameter):
     # See base class.
 
     # object variables
-    # FIXME[typing]: https://github.com/python-attrs/attrs/issues/1197
-    _values: tuple = field(
+    _values: Tuple[str, ...] = field(
         converter=tuple,
-        validator=[min_len(2), validate_unique_values],  # type: ignore
+        validator=(
+            min_len(2),
```
Comment (AVHopp) [inline baybe/parameters/categorical.py:28]: Technically, although a bit weird, there could be a parameter that only has a single value. Do we want to allow that?

Comment (AVHopp) [reply baybe/parameters/categorical.py:28]: Same question for the other classes, e.g.the custom one.

File: baybe/parameters/substance.py [hunk 11]
```diff
@@ -22,6 +23,9 @@
     if _MORDRED_INSTALLED:
         from baybe.utils import smiles_to_mordred_features
 
+Smiles = str
```
Comment (AVHopp) [inline baybe/parameters/substance.py:26]: Having just the first letter capitalized looks weird to me, is this how a type alias should be defined?

File: tests/hypothesis_strategies.py [hunk 12]
```diff
@@ -0,0 +1,158 @@
+"""Hypothesis strategies."""
+import random
+
+import hypothesis.strategies as st
+import numpy as np
+from hypothesis import assume
+from hypothesis.extra.pandas import columns, data_frames
+
+from baybe.exceptions import NumericalUnderflowError
+from baybe.parameters.categorical import (
```
Comment (AVHopp) [inline tests/hypothesis_strategies.py:10]: Isn't it possible to import all of the parameters directly from `baybe.parameters`? Why do we need or want all of the groups here?

File: tests/hypothesis_strategies.py [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+
+_largest_lower_interval = np.nextafter(
+    np.nextafter(np.inf, 0, dtype=DTypeFloatNumpy), 0, dtype=DTypeFloatNumpy
+)
+"""
+The largest possible value for the lower end of a continuous interval such that there
+still exists a larger but finite number for the upper interval end.
+"""
+
+_decorrelate = st.one_of(
```
Comment (AVHopp) [inline tests/hypothesis_strategies.py:31]: Why is this strategy "private"?

File: tests/hypothesis_strategies.py [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+def substance_data(draw: st.DrawFn):
+    """Generate data for class:`baybe.parameters.substance.SubstanceParameter`."""
+    names = draw(st.lists(st.text(min_size=1), min_size=2, max_size=10, unique=True))
+    substances = draw(st.lists(smiles(), min_size=len(names), max_size=len(names)))
+    return dict(zip(names, substances))
+
+
+@st.composite
+def custom_encodings(draw: st.DrawFn):
+    """Generate data for class:`baybe.parameters.custom.CustomDiscreteParameter`."""
```
Comment (AVHopp) [inline tests/hypothesis_strategies.py:65]: I think there is a : missing in front of class. Please check and verify this (and all other similar) link(s).

**Review by AVHopp — APPROVED** [review]
All of my comments are rather minor, so approve :)

File: baybe/parameters/substance.py (hunk 11 above)
Comment (AdrianSosic) [reply baybe/parameters/substance.py:26]: Yup, the convention for aliases is the same as for classes: CamelCase. At least that's my state of knowledge

File: tests/hypothesis_strategies.py (hunk 13 above)
Comment (AdrianSosic) [reply tests/hypothesis_strategies.py:31]: you asked me the same question already on Azure :D go and see if my answer there still makes sense to you. If not, let me know ;)

File: tests/hypothesis_strategies.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply tests/hypothesis_strategies.py:70]: But if that's what he means, then my answer is valid: this is a keyword argument --> there should be no whitespaces :D check literally ALL our remaining code or try to add them! Ruff will happily send them to trash after you've done so

File: baybe/parameters/substance.py (hunk 11 above)
Comment (AVHopp) [reply baybe/parameters/substance.py:26]: Ok, then I'll trust you on this ;)

File: tests/hypothesis_strategies.py [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+    encoding = draw(st.sampled_from(CategoricalEncoding))
+    return CategoricalParameter(name=name, values=values, encoding=encoding)
+
+
+@st.composite
+def task_parameter(draw: st.DrawFn):
+    """Generate class:`baybe.parameters.categorical.TaskParameter`."""
+    name = draw(parameter_name)
+    values = draw(categories)
+    active_values = random.sample(values, random.randint(0, len(values)))
```
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply tests/hypothesis_strategies.py]: @AVHopp: for random.randint, the upper end included:

@Scienfitz: This one is REALLY strange. First: you are right, should be 1. However, the strange part is that the hypothesis test did not create a failing example. I double checked and, indeed, in none of the runs (no matter how large I set the number of examples to be generated), the lower bound of randint is drawn. Even more surprisingly, it **does** draw the lower bound occasionally when you redraw several times. So something really strange is going on with the RNG here. Replacing randint with the numpy version, for example, fixes the error!?!? Anyway, I also realized that using a hypothesis built-in generation has the additional benefit that it allows for shrinkage, so I've replaced that line of code altogether.

Please double-check both and resolve if you are happy. And do let me know if one of you has an explanation for the above!!

[the remark this answers is not in the document]
Comment (AVHopp) [reply tests/hypothesis_strategies.py]: Looks good to me. If @Scienfitz agrees we can resolve imo.

File: tests/hypothesis_strategies.py (hunk 12 above)
Comment (AdrianSosic) [reply tests/hypothesis_strategies.py]: Yes, it is. I've changed it for this PR for the sake of this comment, but we should really have an alignment how we want to handle imports within our own packages (which may be different from what users would do). If we decide this is what we prefer, dann we should probably change this in all files in a separate PR 👍🏼

File: tests/hypothesis_strategies.py (hunk 14 above)
Comment (AdrianSosic) [reply tests/hypothesis_strategies.py]: Absolutely. Damn copy-paste 😄 thanks! Got unnoticed because we are not building the doc for our tests. Could debate if we actually want to do it or not.

Comment (AdrianSosic) [reply tests/hypothesis_strategies.py]: Fixed it, btw

File: baybe/parameters/categorical.py (hunk 10 above)
Comment (AdrianSosic) [reply baybe/parameters/categorical.py:28]: I've asked myself that question already a long time ago (when the parameters file was created) and I think I have a clear answer to it: Yes, having single-valued parameters should be supported, but only via a separate Parameter type, which we could call for instance `FixedParameter`, `StaticParameter`, or `ContextParameter`. The rationale: we want to give users the possibility to explicitly declare the existence of additional single-valued columns in their data (e.g. meta-information of their experiments) and BayBE should swallow those without error. However, these parameter need to be handled differently internally (similar to the TaskParameter), because they would not contribute to the computational representations – their value is fixed, so no need to pass them to any solver.

There is probably already a workitem on Azure, but I can open a PR where we can start working on it. In that same PR, we need to urgently fix a related issue: additional columns are currently simply ignored. We should forbid this and explicitly require either a corresponding regular parameter or "fixed parameter" as described above.

Comment (AdrianSosic) [reply baybe/parameters/categorical.py:28]: TL;DR: no changes required for this PR, but the other PR should follow soon!

File: baybe/exceptions.py [hunk 16]
```diff
@@ -38,3 +38,7 @@ class ModelParamsNotSupportedError(Exception):
 
 class NoRecommendersLeftError(Exception):
     """A recommender is requested by a strategy but there are no recommenders left."""
+
+
+class NumericalUnderflowError(Exception):
+    """A certain computation would lead to numerical underflow."""
```
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply baybe/exceptions.py]: fixed

File: baybe/parameters/custom.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply baybe/parameters/custom.py:72]: I am not aware of the fact that any of our linters would complain about missing entries in the raises section?!? Or is which one do you have in mind?

File: tests/hypothesis_strategies.py (hunk 12 above)
Comment (Scienfitz) [reply tests/hypothesis_strategies.py]: there was a case in a prev PR where importing from the high level caused a circular import problem and only importing from direct source worked so imo thats an argument to import from direct source for use within the package

File: baybe/parameters/categorical.py (hunk 10 above)
Comment (Scienfitz) [reply baybe/parameters/categorical.py:28]: im not sure we should equate metadata with parameters, we have metadata already, seems more straightwordard to me to make this interface proper and use it for metadata rather than inventing new parameters, in a sense this is exactly what its for and you dont have to make filterings for the comp_rep creation

why would you like to disallow additional columns ? This might mean one more unneccessary step for users

File: baybe/parameters/numerical.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Scienfitz) [reply baybe/parameters/numerical.py:62]: imo adding the special case to the valdiator will suffice at the moment

File: tests/hypothesis_strategies.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Scienfitz) [reply tests/hypothesis_strategies.py:94]: im always pro code coverage 
what about creating a common utility that both the test and the implementation use to get the smallest/largest possible tolerance

File: tests/hypothesis_strategies.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Scienfitz) [reply tests/hypothesis_strategies.py:45]: hmm im still undecided. We do not require that SMILES entered are canonical, is is generally not human readable whether a SMILES is canonical or not (so it would be quite a user experience downgrade) and we can work perfectly without canonical ones. However, IF we want to make sure all SMILES describe different molecules the way to go would be via creating the canonical smiles in the backend somewhere to perform the check. Overal thats a really niche issue that can be postponed to another PR. I was just worried that in the small space of molecules that you create here identical ones might be there causing problems down the line due to singularities (eg NCC and CCN)

File: tests/hypothesis_strategies.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Scienfitz) [reply tests/hypothesis_strategies.py:113]: sure since the attrs converter was implemented correctly it would simpy test the list conversion which doesnt seem very exciting. But making exactly sure that it was implemented/included correctly is the point of also testing the string based initialization, who knows what will change in the attrs related lines in the future. Especially for the enums imo thats important as most suers will prob prefer the string based initialization.

im missing both contexts here what is shrinkage in the context of this package (couldnt find it in their docs either) and what the hypothesis strategies are intended to be if youre saying that wouldnt belong here. But if it requires something complicated imo its minor enough to skip it

File: tests/hypothesis_strategies.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (Scienfitz) [reply tests/hypothesis_strategies.py]: yep
(the weird effect adrian describes is concerning tho, does anyone understand RNG?)

File: tests/hypothesis_strategies.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply tests/hypothesis_strategies.py:45]: Update: I'm currently on it and will push a solution in the next minutes. Takeaways:
* The check will only happen in the converter, no need for the user to adjust anything
* Since our `is_valid_smiles` string already called rdkit anyway, there is no real additional load, I think
* I reworked the validator completely such that we now also get nice exceptiongroups (instead of stopping the program when encountering the first problematic input --> should to the same strategy in other validators when we touch then next time). Here a small taster:
  
image

image

[the remark this answers is not in the document]
Comment (AdrianSosic) [reply tests/hypothesis_strategies.py:45]: Now online. If you like it, resolve. Otherwise, let me know :D

File: tests/serialization/test_parameter_serialization.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply tests/serialization/test_parameter_serialization.py:1]: Happy to do so, but this is much easier discussed in a meeting. Can we postpone it till then? I will leave this thread open until we've discussed it ....

File: tests/hypothesis_strategies.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply tests/hypothesis_strategies.py]: Regarding the weird behavior: I looked a bit deeper and think that it was probably due to wrong use of the framework. Since I manually added a different source of randomness that was not under control of hypothesis, the generated scenarios did not really cover the space of possibilities. I've opened an [issue](https://github.com/HypothesisWorks/hypothesis/issues/3810) on their page just to make sure this is what went wrong. So I'll keep track of it from there and resolve here.

File: tests/hypothesis_strategies.py (hunk 12 above)
Comment (AVHopp) [reply tests/hypothesis_strategies.py]: Ok, that works for me. Only in parts that are explicitly pointed towards users (user guides and examples), I'd propose to import stuff like a user would do.

File: tests/hypothesis_strategies.py (hunk 14 above)
Comment (AVHopp) [reply tests/hypothesis_strategies.py]: Is there any reason not to include building the documentation as a test? We do not have broken links towards github pages, and iirc, these were the only ones preventing us from including this, right? If you feel that we might want to discuss this, please open a corresponding discussion (and feel free to resolve here in any case)

File: baybe/parameters/categorical.py (hunk 10 above)
Comment (AVHopp) [reply baybe/parameters/categorical.py:28]: Your explanation makes sense to me, so no reason to change anything here. I think it would be good if you could copy your explanation here to a discussion as this is probably the correct place to discuss the questions and concerns regarding this.

File: baybe/parameters/custom.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (AVHopp) [reply baybe/parameters/custom.py:72]: Huh, seems like I just misremembered this being a thing.

File: tests/hypothesis_strategies.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (AVHopp) [reply tests/hypothesis_strategies.py:45]: Looks good to me, if @Scienfitz agrees we can resolve.

File: baybe/parameters/categorical.py (hunk 10 above)
Comment (AdrianSosic) [reply baybe/parameters/categorical.py:28]: Don't get me wrong, @Scienfitz, I mean metadata in a different sense. Let answer your question first and then it will become clear:

The reason is it because it can give you the absolute worst silent bugs in terms of DOE, and I accidentally stepped into it already twice in my projects and only noticed during debugging! Assume you simply **forget** to declare a parameter. Then, when reading in measurement data that really depends on those parameter's values, you completely neglect the dependence and get a completely wrong model. Chances that this happens when setting up a new project are medium because that's when you think about your data and design space (still, happened to me!) but significantly higher when you adjust your data collection pipeline later on and suddenly have additional relevant columns (happened to me, too!). I think there is absolutely no downside of requiring explicit definition of all columns because the only thing that can happen is that you raise awareness to the user that there is additional stuff floating around that is unspecified. Adding an explicit and simple column selection line of code is really much better than the alternative. Do you agree?

So back to the "metadata" Parameters. Their main purpose would really be a sanity check. If people know that their experiments are all conducted on "machine A" and this information is contained in the data, then this can help avoiding accidentally passing data from "machine B". You know what I mean?

File: tests/hypothesis_strategies.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply tests/hypothesis_strategies.py]: @Scienfitz, @AVHopp, short update: turns out the situation is how I expected it: bug was due to wrong usage of the framework (see issue linked above). So everything's good now, will add one more commit that also removes the other non-built in generators.

File: tests/hypothesis_strategies.py (hunk 12 above)
Comment (AdrianSosic) [reply tests/hypothesis_strategies.py:9]: Ok, then let's keep the policy like that! I've dropped my commit accordingly

File: baybe/parameters/numerical.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply baybe/parameters/numerical.py:62]: done

File: tests/hypothesis_strategies.py (hunk 14 above)
Comment (AdrianSosic) [reply tests/hypothesis_strategies.py]: I didn't mean "building the documentation as a test" but "building the documentation for our test code", which is currently not done since the test code is also not displayed on our github.io page. I was just saying: the error would have been noticed if the test folder was part of the doc generation.... Since it is not, it's also a bit questionable why I've put the class-links there in the first place. So what should we do? 
1) Exclude the test folder from the pipeline and effectively have no check for links / remove the links in the first place 
2) Generate the docs to check the links but not put in on the web
3) Generate the docs to check the links and put it also on the web

File: tests/hypothesis_strategies.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply tests/hypothesis_strategies.py]: Done, the tolerance is now generated as well. Did not do it via utility though because in the parameter class, the necessary sorting step is already done in the attribute converter, so it cannot be extracted nicely into a utility. If you are happy with the code, please resolve 👍🏼

File: tests/hypothesis_strategies.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply tests/hypothesis_strategies.py:122]: Ok, I see your point. Even though chances are really really small that this will break in the future, adding a short test case doesn't hurt either, so I've added them. For now, I went via two additional explicit tests instead of extending the strategy, as described above. We can still discuss in our meeting for what we intend to use hypothesis and for what not, and we can change it after the meeting if necessary. For now, I went with the credo "only span the space of possibilities". Please resolve if this is OK

File: tests/serialization/test_parameter_serialization.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Scienfitz) [reply tests/serialization/test_parameter_serialization.py:1]: yes lets discuss in the meeting, for such drastic changes imo that should prob even happen before implementation starts

File: baybe/parameters/categorical.py (hunk 10 above)
Comment (Scienfitz) [reply baybe/parameters/categorical.py:28]: imo this should be moved to github discussions and thread can be closed
as is, theres no parameter that should be allowed a singular value

Comment (AdrianSosic) [reply baybe/parameters/categorical.py:28]: Done, see [here](https://github.com/emdgroup/baybe/discussions/62). Will resolve

File: tests/hypothesis_strategies.py (hunk 14 above)
Comment (AdrianSosic) [reply tests/hypothesis_strategies.py]: Moved this to a [separate discussion](https://github.com/emdgroup/baybe/discussions/63) since not related to the PR. Will resolve

**Review by Scienfitz — APPROVED** [review]
